File: shared/python/contracts/direction_v3_1.py
"""Direction Format v3.1 — Pydantic mirror of ``directionV3_1.ts``.

Mirrors the Zod schema in ``backend/media/remotion/src/schemas/directionV3_1.ts``.
Any change here must be reflected there and vice-versa. See
``/Users/saurabhrawat/.devin/plans/plan-476749920bd78127.md`` for the rationale.

Usage:

    from contracts.direction_v3_1 import DirectionV3_1, validate_timeline_density

    doc = DirectionV3_1.model_validate(payload)
    issues = validate_timeline_density(doc)
    if issues:
        logger.warning("direction.density_gaps", issues=issues)

Design notes:
- All v3.1-added fields are ``Optional`` so a v3.0 payload parses unchanged.
- ``timeline[]`` keyframes are anchor points; between them Remotion
  interpolates using the ``ease`` field. Density floor is ``MAX_KEYFRAME_GAP_MS``.
- ``captions[]`` mirror voice's ``word_alignment`` (from Inworld TTS
  ``get_word_timestamps`` or whisperx fallback).
- ``micro_beats[]`` are ≤100ms hit markers merged from voice ``emphasis_hits``
  + music ``beat_map_ms`` + asset ``cut_suggestions_ms``.
"""

from __future__ import annotations

from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field, HttpUrl

MAX_KEYFRAME_GAP_MS = 500


# ── enums ──────────────────────────────────────────────────────────────────


class EaseCurve(str, Enum):
    LINEAR = "linear"
    STEP = "step"
    CUBIC_IN = "cubic-in"
    CUBIC_OUT = "cubic-out"
    CUBIC_IN_OUT = "cubic-in-out"
    SPRING = "spring"
    ELASTIC_OUT = "elastic-out"
    BACK_OUT = "back-out"
    EXPO_OUT = "expo-out"


class MicroBeatKind(str, Enum):
    HIT = "hit"
    CUT = "cut"
    REVEAL = "reveal"
    ZOOM_PUNCH = "zoom_punch"
    FLASH = "flash"
    SFX = "sfx"


class MicroBeatSource(str, Enum):
    VOICE_EMPHASIS = "voice_emphasis"
    MUSIC_BEAT = "music_beat"
    ASSET_CUT = "asset_cut"
    NARRATIVE = "narrative"
    MANUAL = "manual"


class Aspect(str, Enum):
    WIDE = "16:9"
    VERTICAL = "9:16"
    SQUARE = "1:1"


class LayerFit(str, Enum):
    COVER = "cover"
    CONTAIN = "contain"
    FILL = "fill"


class DirectionVersion(str, Enum):
    V3_0 = "3.0"
    V3_1 = "3.1"


# ── shared primitives (v3.0 kept) ──────────────────────────────────────────


class PresetRef(BaseModel):
    model_config = ConfigDict(extra="allow")
    preset: str = Field(..., min_length=1)
    overrides: Optional[dict[str, Any]] = None


class Animation(BaseModel):
    model_config = ConfigDict(extra="allow")
    preset: str = Field(..., min_length=1)
    target: Optional[str] = None
    overrides: Optional[dict[str, Any]] = None


class SfxCue(BaseModel):
    preset: str
    at_ms: int = Field(..., ge=0)
    volume_db: Optional[float] = None


# ── v3.1 additions ─────────────────────────────────────────────────────────


class CameraState(BaseModel):
    x: float = 0.0
    y: float = 0.0
    scale: float = Field(1.0, gt=0)
    rotation: float = 0.0
    anchor: tuple[float, float] = (0.5, 0.5)


class EffectsIntensity(BaseModel):
    grade_strength: float = Field(1.0, ge=0, le=1)
    vignette: float = Field(0.0, ge=0, le=1)
    blur_px: float = Field(0.0, ge=0)
    chroma: float = Field(0.0, ge=0, le=1)
    grain: float = Field(0.0, ge=0, le=1)
    bloom: float = Field(0.0, ge=0, le=1)


class LightingState(BaseModel):
    exposure: float = 0.0
    contrast: float = Field(1.0, gt=0)
    saturation: float = Field(1.0, ge=0)
    temperature: float = 0.0
    tint: float = 0.0


class TextState(BaseModel):
    visible_word_index: int = Field(0, ge=0)
    emphasis_word_index: Optional[int] = Field(None, ge=0)
    opacity: float = Field(1.0, ge=0, le=1)
    scale: float = Field(1.0, gt=0)


class OverlayOpacity(BaseModel):
    lower_third: float = Field(0.0, ge=0, le=1)
    watermark: float = Field(1.0, ge=0, le=1)
    caption: float = Field(1.0, ge=0, le=1)
    progress_bar: float = Field(0.0, ge=0, le=1)


class TimelineKeyframe(BaseModel):
    """Anchor keyframe. ≤500ms spacing required (see `validate_timeline_density`)."""

    model_config = ConfigDict(extra="allow")
    t_ms: int = Field(..., ge=0)
    ease: EaseCurve = EaseCurve.CUBIC_IN_OUT
    camera: Optional[CameraState] = None
    effects: Optional[EffectsIntensity] = None
    lighting: Optional[LightingState] = None
    text_state: Optional[TextState] = None
    overlay_opacity: Optional[OverlayOpacity] = None
    audio_ducking_db: Optional[float] = None


class CaptionWord(BaseModel):
    """Word-level caption from voice's `word_alignment` (Inworld or whisperx)."""

    word: str
    start_ms: int = Field(..., ge=0)
    end_ms: int = Field(..., gt=0)
    is_emphasis: bool = False
    sentence_id: Optional[str] = None
    style_id: Optional[str] = None


class MicroBeat(BaseModel):
    """Sub-100ms hit marker. Merged from voice / music / asset sources."""

    at_ms: int = Field(..., ge=0)
    kind: MicroBeatKind
    source: MicroBeatSource
    intensity: float = Field(0.5, ge=0, le=1)
    payload: Optional[dict[str, Any]] = None


class DuckingEnvelopePoint(BaseModel):
    t_ms: int = Field(..., ge=0)
    gain_db: float


class AudioTrack(BaseModel):
    voiceover_url: Optional[HttpUrl] = None
    voiceover_start_ms: int = Field(0, ge=0)
    voiceover_gain_db: float = 0.0
    music_bed_url: Optional[HttpUrl] = None
    music_gain_db: float = -8.0
    ducking_envelope: Optional[list[DuckingEnvelopePoint]] = None
    loudness_target_lufs: float = -14.0


class LayerOpacityKeyframe(BaseModel):
    t_ms: int = Field(..., ge=0)
    opacity: float = Field(..., ge=0, le=1)


class Layer(BaseModel):
    z: int = 0
    asset_url: HttpUrl
    fit: LayerFit = LayerFit.COVER
    transform_keyframes: Optional[list[TimelineKeyframe]] = None
    opacity_keyframes: Optional[list[LayerOpacityKeyframe]] = None


# ── segment (v3.0 fields kept + v3.1 additions) ───────────────────────────


class Segment(BaseModel):
    model_config = ConfigDict(extra="allow")

    # v3.0 (unchanged)
    id: str
    start_ms: int = Field(..., ge=0)
    duration_ms: int = Field(..., gt=0)
    scene_preset: str = Field(..., min_length=1)
    scene_overrides: Optional[dict[str, Any]] = None
    animations_in: Optional[list[Animation]] = None
    animations_out: Optional[list[Animation]] = None
    effects: Optional[list[str]] = None
    overlays: Optional[list[PresetRef]] = None
    sfx: Optional[list[SfxCue]] = None
    transition_out: Optional[PresetRef] = None

    # v3.1 additions
    timeline: Optional[list[TimelineKeyframe]] = None
    captions: Optional[list[CaptionWord]] = None
    micro_beats: Optional[list[MicroBeat]] = None
    audio_track: Optional[AudioTrack] = None
    layers: Optional[list[Layer]] = None
    is_hero_moment: Optional[bool] = None


# ── top-level (v3.0 fields kept) ──────────────────────────────────────────


class Theme(BaseModel):
    primary_color: str
    accent_color: str
    background_color: str
    text_color: str
    fonts: dict[str, str]


class DuckingConfig(BaseModel):
    enabled: bool
    threshold_db: float
    attack_ms: float
    release_ms: float


class Audio(BaseModel):
    voiceover_url: Optional[HttpUrl] = None
    voiceover_srt_url: Optional[HttpUrl] = None
    music_url: Optional[HttpUrl] = None
    music_volume_db: Optional[float] = None
    ducking: Optional[DuckingConfig] = None
    loudness_target_lufs: Optional[float] = None


class Branding(BaseModel):
    intro_preset: Optional[str] = None
    outro_preset: Optional[str] = None
    watermark: Optional[PresetRef] = None


class Resolution(BaseModel):
    width: int = Field(..., gt=0)
    height: int = Field(..., gt=0)


class Meta(BaseModel):
    video_id: str
    channel_id: str
    title: str
    duration_target_seconds: float = Field(..., gt=0)
    aspect: Aspect
    fps: int = Field(..., gt=0)
    resolution: Resolution


class ThumbnailTitle(BaseModel):
    text: str
    font: str
    weight: int
    size: int
    color: str
    stroke: Optional[dict[str, Any]] = None


class Thumbnail(BaseModel):
    composition: str  # "ThumbnailComp"
    layout_preset: str
    background_url: HttpUrl
    title: ThumbnailTitle
    format: str  # "png" | "jpg"
    width: int = Field(..., gt=0)
    height: int = Field(..., gt=0)


class DirectionV3_1(BaseModel):
    """Root Direction v3.1 payload. Also accepts v3.0 (all v3.1 fields optional)."""

    model_config = ConfigDict(extra="allow")

    version: DirectionVersion
    meta: Meta
    template: str
    theme: Theme
    grade_preset: str
    global_effects: Optional[list[str]] = None
    global_overlays: Optional[list[PresetRef]] = None
    audio: Optional[Audio] = None
    branding: Optional[Branding] = None
    segments: list[Segment] = Field(..., min_length=1)
    thumbnail: Optional[Thumbnail] = None


# ── density validator ─────────────────────────────────────────────────────


def validate_timeline_density(direction: DirectionV3_1) -> list[str]:
    """Check every segment's timeline has ≤500 ms gaps.

    Only enforced when a segment actually declares a ``timeline`` — v3.0
    payloads (no timeline) pass unchecked. Assembly service upgrades
    v3.0 → v3.1 via the densifier before this check.

    Returns a list of human-readable issues. Empty list = passes.
    """
    issues: list[str] = []
    for seg in direction.segments:
        if not seg.timeline:
            continue
        kfs = sorted(seg.timeline, key=lambda k: k.t_ms)
        if kfs[0].t_ms > 0:
            issues.append(f"segment {seg.id}: first keyframe at {kfs[0].t_ms}ms, expected 0")
        last = kfs[-1]
        tail_gap = seg.duration_ms - last.t_ms
        if tail_gap > MAX_KEYFRAME_GAP_MS:
            issues.append(
                f"segment {seg.id}: last keyframe at {last.t_ms}ms leaves "
                f"{tail_gap}ms tail (> {MAX_KEYFRAME_GAP_MS}ms floor) before segment end"
            )
        for i in range(1, len(kfs)):
            gap = kfs[i].t_ms - kfs[i - 1].t_ms
            if gap > MAX_KEYFRAME_GAP_MS:
                issues.append(
                    f"segment {seg.id}: keyframe gap {gap}ms between "
                    f"t={kfs[i - 1].t_ms} and t={kfs[i].t_ms} exceeds {MAX_KEYFRAME_GAP_MS}ms floor"
                )
    return issues

File: backend/media/remotion/src/schemas/directionV3_1.ts
import { z } from "zod";

/**
 * Direction Format v3.1 — granular editorial contract.
 *
 * v3.1 extends v3.0 with the granularity needed for professional-tier editing:
 *   - `timeline[]` per segment — anchor keyframes at ≤500ms spacing (½-sec floor).
 *     Each keyframe carries full editorial state (camera, effects, lighting,
 *     text visibility, overlay opacity, audio ducking). Between keyframes,
 *     Remotion interpolates via the `ease` field on each keyframe.
 *   - `captions[]` per segment — word-level timing (from Inworld TTS
 *     word_alignment or whisperx fallback). Enables word-perfect subtitle
 *     rendering and per-word emphasis animations.
 *   - `micro_beats[]` per segment — sub-100ms hit markers merged from voice
 *     emphasis_hits + music beat_map_ms + asset cut_suggestions_ms.
 *     Fires punch-zooms, flashes, cuts on exact frames.
 *   - `audio_track` per segment — measured voice URL/gain, music URL/gain,
 *     and a per-frame ducking_envelope so music dips exactly during speech.
 *   - `layers[]` per segment — parallax / PIP / secondary video stacks.
 *
 * Backward compatibility: all new fields are optional. A v3.0 payload
 * parses under v3.1 unchanged. The `direction` service upgrades v3.0 →
 * v3.1 in-place via the densifier described in the plan.
 *
 * Rule (enforced by validator, not by Zod alone):
 *   for every segment, `timeline` covers [0, duration_ms] with each pair
 *   of consecutive `t_ms` values ≤ 500 ms apart (the ½-second floor).
 *   `assembly` service will auto-interpolate + raise a sync_issues warning
 *   when the input is sparser than that.
 */

const presetId = z.string().min(1);

const PresetRef = z.object({
  preset: presetId,
  overrides: z.record(z.unknown()).optional(),
});

const Animation = z.object({
  preset: presetId,
  target: z.string().optional(),
  overrides: z.record(z.unknown()).optional(),
});

const SfxCue = z.object({
  preset: z.string(),
  at_ms: z.number().int().nonnegative(),
  volume_db: z.number().optional(),
});

// ── NEW in v3.1 ────────────────────────────────────────────────────────────

const EaseCurve = z.enum([
  "linear",
  "step",
  "cubic-in",
  "cubic-out",
  "cubic-in-out",
  "spring",
  "elastic-out",
  "back-out",
  "expo-out",
]);

const CameraState = z.object({
  x: z.number().default(0), // px, relative to composition center
  y: z.number().default(0),
  scale: z.number().positive().default(1), // 1.0 = fit; 1.15 = 15% zoom in
  rotation: z.number().default(0), // degrees
  anchor: z.tuple([z.number(), z.number()]).default([0.5, 0.5]), // normalized [0..1] pivot
});

const EffectsIntensity = z.object({
  grade_strength: z.number().min(0).max(1).default(1), // color-grade LUT mix amount
  vignette: z.number().min(0).max(1).default(0),
  blur_px: z.number().min(0).default(0),
  chroma: z.number().min(0).max(1).default(0), // chromatic aberration
  grain: z.number().min(0).max(1).default(0),
  bloom: z.number().min(0).max(1).default(0),
});

const LightingState = z.object({
  exposure: z.number().default(0), // stops, ±3
  contrast: z.number().positive().default(1),
  saturation: z.number().nonnegative().default(1),
  temperature: z.number().default(0), // -100..+100 (cool→warm)
  tint: z.number().default(0), // green↔magenta
});

const TextState = z.object({
  visible_word_index: z.number().int().nonnegative().default(0), // reveal index for typewriter/karaoke
  emphasis_word_index: z.number().int().nonnegative().nullable().default(null),
  opacity: z.number().min(0).max(1).default(1),
  scale: z.number().positive().default(1),
});

const OverlayOpacity = z.object({
  lower_third: z.number().min(0).max(1).default(0),
  watermark: z.number().min(0).max(1).default(1),
  caption: z.number().min(0).max(1).default(1),
  progress_bar: z.number().min(0).max(1).default(0),
});

/**
 * Anchor keyframe. At least one keyframe per ½-second is required (assembly
 * enforces). Values here are what the state SHOULD BE at `t_ms`; between
 * consecutive keyframes Remotion interpolates using the `ease` on the
 * later keyframe.
 */
const TimelineKeyframe = z.object({
  t_ms: z.number().int().nonnegative(), // absolute ms within the segment
  ease: EaseCurve.default("cubic-in-out"),
  camera: CameraState.partial().optional(),
  effects: EffectsIntensity.partial().optional(),
  lighting: LightingState.partial().optional(),
  text_state: TextState.partial().optional(),
  overlay_opacity: OverlayOpacity.partial().optional(),
  audio_ducking_db: z.number().optional(), // 0 = no duck, -12 = duck music by 12 dB
});

const CaptionWord = z.object({
  word: z.string(),
  start_ms: z.number().int().nonnegative(),
  end_ms: z.number().int().positive(),
  is_emphasis: z.boolean().default(false),
  sentence_id: z.string().optional(),
  style_id: z.string().optional(), // preset ID for this word's caption style
});

const MicroBeatKind = z.enum([
  "hit", // impact — flash, shake, thud
  "cut", // scene change
  "reveal", // text/overlay appears
  "zoom_punch", // rapid zoom-in
  "flash", // 1–2 frame full-screen flash
  "sfx", // trigger an SFX
]);

const MicroBeat = z.object({
  at_ms: z.number().int().nonnegative(),
  kind: MicroBeatKind,
  source: z.enum(["voice_emphasis", "music_beat", "asset_cut", "narrative", "manual"]),
  intensity: z.number().min(0).max(1).default(0.5),
  payload: z.record(z.unknown()).optional(), // kind-specific extras
});

const DuckingEnvelopePoint = z.object({
  t_ms: z.number().int().nonnegative(),
  gain_db: z.number(), // e.g. 0 during silence, -12 during speech
});

const AudioTrack = z.object({
  voiceover_url: z.string().url().optional(),
  voiceover_start_ms: z.number().int().nonnegative().default(0),
  voiceover_gain_db: z.number().default(0),
  music_bed_url: z.string().url().optional(),
  music_gain_db: z.number().default(-8), // typical music bed sits ~8 dB below voice
  ducking_envelope: z.array(DuckingEnvelopePoint).optional(),
  loudness_target_lufs: z.number().default(-14), // YouTube target
});

const Layer = z.object({
  z: z.number().int().default(0), // stack order — higher = on top
  asset_url: z.string().url(),
  fit: z.enum(["cover", "contain", "fill"]).default("cover"),
  transform_keyframes: z
    .array(TimelineKeyframe.partial().extend({ t_ms: z.number().int().nonnegative() }))
    .optional(),
  opacity_keyframes: z
    .array(z.object({ t_ms: z.number().int().nonnegative(), opacity: z.number().min(0).max(1) }))
    .optional(),
});

// ── Segment (v3.0 fields kept + v3.1 fields added) ─────────────────────────

const Segment = z.object({
  // v3.0 (unchanged)
  id: z.string(),
  start_ms: z.number().int().nonnegative(),
  duration_ms: z.number().int().positive(),
  scene_preset: presetId,
  scene_overrides: z.record(z.unknown()).optional(),
  animations_in: z.array(Animation).optional(),
  animations_out: z.array(Animation).optional(),
  effects: z.array(presetId).optional(),
  overlays: z.array(PresetRef).optional(),
  sfx: z.array(SfxCue).optional(),
  transition_out: PresetRef.optional(),

  // v3.1 additions (all optional for backward-compat)
  timeline: z.array(TimelineKeyframe).optional(),
  captions: z.array(CaptionWord).optional(),
  micro_beats: z.array(MicroBeat).optional(),
  audio_track: AudioTrack.optional(),
  layers: z.array(Layer).optional(),
  is_hero_moment: z.boolean().optional(), // hook / punchline / reveal → gets multi-take voice
});

const Theme = z.object({
  primary_color: z.string(),
  accent_color: z.string(),
  background_color: z.string(),
  text_color: z.string(),
  fonts: z.object({ heading: z.string(), body: z.string() }),
});

const Audio = z.object({
  voiceover_url: z.string().url().optional(),
  voiceover_srt_url: z.string().url().optional(),
  music_url: z.string().url().optional(),
  music_volume_db: z.number().optional(),
  ducking: z
    .object({
      enabled: z.boolean(),
      threshold_db: z.number(),
      attack_ms: z.number(),
      release_ms: z.number(),
    })
    .optional(),
  loudness_target_lufs: z.number().optional(),
});

const Branding = z
  .object({
    intro_preset: z.string().optional(),
    outro_preset: z.string().optional(),
    watermark: PresetRef.optional(),
  })
  .optional();

const Meta = z.object({
  video_id: z.string(),
  channel_id: z.string(),
  title: z.string(),
  duration_target_seconds: z.number().positive(),
  aspect: z.enum(["16:9", "9:16", "1:1"]),
  fps: z.number().int().positive(),
  resolution: z.object({
    width: z.number().int().positive(),
    height: z.number().int().positive(),
  }),
});

const Thumbnail = z
  .object({
    composition: z.literal("ThumbnailComp"),
    layout_preset: z.string(),
    background_url: z.string().url(),
    title: z.object({
      text: z.string(),
      font: z.string(),
      weight: z.number(),
      size: z.number(),
      color: z.string(),
      stroke: z.object({ color: z.string(), width: z.number() }).optional(),
    }),
    format: z.enum(["png", "jpg"]),
    width: z.number().int().positive(),
    height: z.number().int().positive(),
  })
  .optional();

export const DirectionV3_1 = z.object({
  version: z.union([z.literal("3.0"), z.literal("3.1")]), // accept both
  meta: Meta,
  template: z.string(),
  theme: Theme,
  grade_preset: presetId,
  global_effects: z.array(presetId).optional(),
  global_overlays: z.array(PresetRef).optional(),
  audio: Audio.optional(),
  branding: Branding,
  segments: z.array(Segment).min(1),
  thumbnail: Thumbnail,
});

export type DirectionV3_1Input = z.infer<typeof DirectionV3_1>;
export type SegmentV3_1 = z.infer<typeof Segment>;
export type TimelineKeyframeInput = z.infer<typeof TimelineKeyframe>;
export type CaptionWordInput = z.infer<typeof CaptionWord>;
export type MicroBeatInput = z.infer<typeof MicroBeat>;
export type AudioTrackInput = z.infer<typeof AudioTrack>;
export type LayerInput = z.infer<typeof Layer>;

/**
 * Runtime check: every segment's `timeline` must cover [0, duration_ms]
 * with consecutive keyframes ≤ 500 ms apart. Returns a list of issues.
 * Empty list = passes. Call this in `assembly` after densification.
 */
export const MAX_KEYFRAME_GAP_MS = 500;

export function validateTimelineDensity(direction: DirectionV3_1Input): string[] {
  const issues: string[] = [];
  for (const seg of direction.segments) {
    if (!seg.timeline || seg.timeline.length === 0) {
      // v3.1 permits omission — density is only checked when timeline is present.
      continue;
    }
    const kfs = [...seg.timeline].sort((a, b) => a.t_ms - b.t_ms);
    const first = kfs[0];
    const last = kfs[kfs.length - 1];
    if (!first || !last) continue;
    if (first.t_ms > 0) {
      issues.push(`segment ${seg.id}: first keyframe at ${first.t_ms}ms, expected 0`);
    }
    if (last.t_ms < seg.duration_ms - MAX_KEYFRAME_GAP_MS) {
      issues.push(
        `segment ${seg.id}: last keyframe at ${last.t_ms}ms leaves > ${MAX_KEYFRAME_GAP_MS}ms tail before segment end (${seg.duration_ms}ms)`,
      );
    }
    for (let i = 1; i < kfs.length; i++) {
      const prev = kfs[i - 1];
      const curr = kfs[i];
      if (!prev || !curr) continue;
      const gap = curr.t_ms - prev.t_ms;
      if (gap > MAX_KEYFRAME_GAP_MS) {
        issues.push(
          `segment ${seg.id}: keyframe gap ${gap}ms between t=${prev.t_ms} and t=${curr.t_ms} exceeds ${MAX_KEYFRAME_GAP_MS}ms floor`,
        );
      }
    }
  }
  return issues;
}
